Updated efficiency & implemented Notification.Builder

fean opened the pull request:
- All-round efficiency update for Android.
- Implemented the new Notification.Builder class to build the
  notifications for Android devices.

[Warning: The Notification icon is still set to one absolute to my project environment]

fean committed:
Updated efficiency & implemented Notification.Builder

- All-round efficiency update
- Implemented the new Notification.Builder class to build the
notification.

aogilvie commented:
@fean Thanks! It's a national holiday here in Japan, so i'll take a look after the long weekend! Thanks again :+1:

fean commented:
@aogilvie Not a problem, go have some fun!

aogilvie commented on android/jp/wizcorp/phonegap/plugin/localNotification/AlarmReceiver.java:
```diff
@@ -32,31 +33,29 @@
 		}
+		
+		Notification notification = new Notification.Builder(context)
+			.setSmallIcon(org.authiq.emerge.R.drawable.notification)
```
Suggest fetching this on the fly with;

```
int appResId = cordova.getActivity().getResources().getIdentifier("notificationIcon", "drawable", cordova.getActivity().getPackageName());
```

Would require a default if not found and an update for users in the README.mb on what to call their image if they want to add a custom icon.

My personal choice of default: `R.drawable.btn_star_big_on` :star:

aogilvie commented on android/jp/wizcorp/phonegap/plugin/localNotification/LocalNotification.java:
```diff
@@ -57,25 +48,20 @@ public boolean execute(String action, JSONArray args, CallbackContext callbackCo
 			    persistAlarm(alarmId, args);
-			    return this.add(callbackContext, title, subTitle, ticker, "" + alarmId, seconds);
+			    return this.add(callbackContext, args.getJSONObject(1).getString("title"),
+			    		args.getJSONObject(1).getString("message"), args.getJSONObject(1).getString("ticker"),
```
`ticker` and `title` are now required fields this should be mentioned in the README.md, perhaps a minor addition to the code sample?

What do you think about adding suitable defaults for some of these? As I understand `ticker` and `message` are very similar in behaviour (i.e. most likely these will have the same string message value)?

aogilvie commented:
@fean Had a quick look through. I have suggested some minor alterations above.

fean commented:
Indeed you're right. I'll add the defaults and set-up the readme.md to comply with the changes.

When a message is put on the statusbar it shows the ticker text and the notification icon for a second. In the case of my project it said: New eMerge notification.

fean committed:
Updated Readme.md

- Changed the whole layout and added descriptions
- Added the new Android only properties to the readme.md

fean committed:
Typo correction

fean committed:
Update README.md

fean committed:
Update

- Added a default value for title, ticker & icon

fean commented:
Ok, the title, ticker and icon property on the JavaScript `add()` function have now become optionals. I've also re-written the readme.md to comply with the changes I made to the API.

fean committed:
Update

- Removed debug info
- Exception handling added

aogilvie commented:
Awesome! Thanks for this.

Two slight indentation issues that would be really quick to fix:
https://github.com/fean/phonegap-plugin-localNotifications/blob/v3.0/android/jp/wizcorp/phonegap/plugin/localNotification/AlarmHelper.java#L47
Line 72 in the readme in the add notification options Object example:
https://github.com/fean/phonegap-plugin-localNotifications/blob/v3.0/README.md

Then this is ready to go in! Thanks!

fean committed:
Indentation issue fixed

fean committed:
Fixed strange indentation

fean commented:
Allright, I just fixed the indentation issues. I think we're good to go.

aogilvie commented on android/jp/wizcorp/phonegap/plugin/localNotification/AlarmHelper.java:
```diff
@@ -33,26 +35,21 @@ public AlarmHelper(Context context) {
+			.putExtra(AlarmReceiver.TITLE, alarmTitle)
+			.putExtra(AlarmReceiver.SUBTITLE, alarmSubTitle)
+			.putExtra(AlarmReceiver.TICKER_TEXT, alarmTicker)
+            		.putExtra(AlarmReceiver.ICON, icon)
```
I double checked this code with the example project and it would not compile because AlarmRestoreOnBoot.java does not pass an `icon` to `addAlarm( )` https://github.com/Wizcorp/phonegap-plugin-localNotifications/blob/v3.0/android/jp/wizcorp/phonegap/plugin/localNotification/AlarmRestoreOnBoot.java#L51

aogilvie commented on android/jp/wizcorp/phonegap/plugin/localNotification/AlarmReceiver.java:
```diff
@@ -32,31 +33,29 @@
+			.setContentIntent(contentIntent)
+			.setVibrate(new long[] { 0, 100, 200, 300 })
+			.setWhen(System.currentTimeMillis())
+			.build();
```
Are you aware that adding `build()` requires a dependency - android-support-v4.jar? I cannot currently compile the example project with your source. :panda_face:

fean commented:
I totally forgot to implement `icon` in  the `AlarmRestoreOnBoot.onReceive()` function. I'll go implement it right away. 

To use the `Notification.Builder.build()` function android-support-v4.jar is not necessary. If you want to ensure code compatibility below Api level 5 you can use the classes in android-support-v4.jar. The notification class you are talking about which resides within the android-support-v4.jar is `NotificationCompat.Builder`. If you'd want to use it's `build()` function then you'd indeed need android-support-v4.jar. 

The android documentation on `Notification.Builder`:

http://developer.android.com/reference/android/app/Notification.Builder.html

fean committed:
Update

- AlarmRestoreOnReboot fixed
- AlarmOptions fixed

aogilvie commented:
OK I ran this with the example project and it works nicely now. Thanks.

The `build()` compile error was in fact caused by the project.properties `target=android-14`, which I changed to 16 (my AndroidManifest was already min 16).

Because this is a breaking change for some users supporting < 16, i'm tagging the current v3.0 as 1.0.0 release. I'll merge your code and update the plugin.xml as 2.0.0 and tag.

Thanks for the update!
